### ISSUE TITLE
`nn.Linear` allows 1d input tensors

### ISSUE THREAD

**Issue by jatentaki** [pytorch/pytorch#16896]
## 🐛 Bug

The [docs](https://pytorch.org/docs/master/nn.html#torch.nn.Linear) for `nn.Linear` claim that the input has to be `[N, *, M]`, but it accepts `[M]` as well. This is because `nn.Linear` dispatches (via `functional.linear`) to `addmm` and `matmul`, which perform this kind of broadcasting. It's unclear if it's a bug or a feature, but certainly needs adapting either the code or the docs.

This report is a follow-up to my [answer](https://stackoverflow.com/q/54591124/4280242) on StackOverflow, which traces the root problem.

## To Reproduce
```python
import torch
import torch.nn as nn

n = 5
lin = nn.Linear(n, n)
inp = torch.randn(n)

print(lin(inp))
```

## Expected behavior

Error stating that 2d input was expected and 1d was received.

## Environment

PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] Could not collect
[conda] pytorch-cpu               1.0.0               py3.6_cpu_1    pytorch
[conda] torch-dimcheck            0.0.1                     <pip>
[conda] torchvision               0.2.1                     <pip>
[conda] torchvision-cpu           0.2.1                    py36_1    pytorch

**Comment by apaszke** [comment]
I'd say it's somewhere between a bug and a feature. We never say that we will check all the boundary conditions explicitly in those functions, so if it ends up working for a different dim configuration that's ok. But you shouldn't treat that as a guarantee that it will continue to work in the future.

**Comment by ngimel** [comment]
1d inputs are still accepted, marked it as a doc issue to update the docs
